modules/sor/sorV2/lib/pathGraph/pathGraph.ts

```diff
@@ -457,38 +506,74 @@ export class PathGraph {
-                SwapKind.GivenIn,
+    private getLimitAmountSwapForPath(path: PathGraphEdgeData[]): bigint {
+        let limit;
+        if (this.getOperationFromPath(path[path.length - 1]) === PathOperation.AddLiquidity) {
+            limit = MAX_UINT256;
+        } else if (this.getOperationFromPath(path[path.length - 1]) === PathOperation.RemoveLiquidity) {
+            limit = path[path.length - 1].pool.getLimitAmountRemoveLiquidity(
+                path[path.length - 1].tokenIn,
+                path[path.length - 1].tokenOut,
+                RemoveLiquidityKind.SingleTokenExactIn,
             );
-            const poolLimitExactOut = path[i].pool.getLimitAmountSwap(
-                path[i].tokenIn,
-                path[i].tokenOut,
-                SwapKind.GivenOut,
+        } else {
+            limit = path[path.length - 1].pool.getLimitAmountSwap(
+                path[path.length - 1].tokenIn,
+                path[path.length - 1].tokenOut,
+                SwapKind.GivenIn,
             );
+        }
 
-            if (poolLimitExactOut <= limit) {
-                limit = poolLimitExactIn;
-            } else {
-                const pulledLimit = path[i].pool.swapGivenOut(
+        for (let i = path.length - 2; i >= 0; i--) {
+            let limitGivenIn;
+            let limitGivenOut;
+            if (this.getOperationFromPath(path[i]) === PathOperation.Swap) {
+                limitGivenIn = path[i].pool.getLimitAmountSwap(path[i].tokenIn, path[i].tokenOut, SwapKind.GivenIn);
+                limitGivenOut = path[i].pool.getLimitAmountSwap(path[i].tokenIn, path[i].tokenOut, SwapKind.GivenOut);
+                if (limitGivenOut <= limit) {
+                    limit = limitGivenOut;
+                } else {
+                    const pulledLimit: bigint = path[i].pool.swapGivenOut(
+                        path[i].tokenIn,
+                        path[i].tokenOut,
+                        TokenAmount.fromRawAmount(path[i].tokenOut, limit),
+                    ).amount;
+                    limit = pulledLimit > limitGivenIn ? limitGivenIn : pulledLimit;
+                }
+            } else if (this.getOperationFromPath(path[i]) === PathOperation.RemoveLiquidity) {
+                limitGivenIn = path[i].pool.getLimitAmountRemoveLiquidity(
                     path[i].tokenIn,
                     path[i].tokenOut,
-                    TokenAmount.fromRawAmount(path[i].tokenOut, limit),
-                ).amount;
-
-                limit = pulledLimit > poolLimitExactIn ? poolLimitExactIn : pulledLimit;
+                    RemoveLiquidityKind.SingleTokenExactIn,
+                );
+                limitGivenOut = path[i].pool.getLimitAmountRemoveLiquidity(
+                    path[i].tokenIn,
+                    path[i].tokenOut,
+                    RemoveLiquidityKind.SingleTokenExactOut,
+                );
+                if (limitGivenOut <= limit) {
+                    limit = limitGivenOut;
```

Comment:
Should be `given in` here as well.
```suggestion
                    limit = limitGivenIn;
```